src/server/player_sao.cpp

```diff
@@ -491,6 +499,12 @@ void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason)
 	// Update properties on death
 	if ((hp == 0) != (oldhp == 0))
 		m_properties_sent = false;
+
+	// Send normal HP updates instantly, but delay death messages
+	if (hp > 0)
+		m_env->getGameDef()->SendPlayerHPOrDie(this, reason);
+	else if (!m_death_reason)
+		m_death_reason = new PlayerHPChangeReason(reason);
```

Comment:
What if there won't be a next step because the player disconnected? Looks like a memory leak

Reply:
PlayerHPChangeReason contains lua_reference. Should ref it there

Reply:
`lua_reference` invalidated in either `l_set_hp` or `step()`, depending on what's called first.